### PR TITLE
Free drift dynamics: now with dynamics switcher!

### PR DESCRIPTION
**Pull request by timspainNERSC**
# Free drift dynamics
## Fixes \#453

---

# Change Description

Add a free drift equivalent to MEVPStep and BBMStep. Running this depends on implementing the dynamics switcher (closing #378).

---

# Test Description

Not yet. Should be tested once this form of dynamics can be tested.

---

# Documentation Impact

The free drift dynamics step needs to be documented along with the other available dynamics steps.

### CODE REVIEW

**Review by MarionBWeinzierl — COMMENTED** [review]
There is quite a bit copy-paste between the different Dynamics implementation (including the same comments about spherical coordinates). I am wondering whether this could be avoided by moving them into a parent class, or otherwise a shared function?

**Comment by timspainNERSC** [comment]
I have completely stomped on the code in the old branch, and cherry picked what I think is necessary to get free-drift dynamics.
@MarionBWeinzierl could you review the code, now it is free of all the repetition?
@einola could you review the maths to make sure I got all the commits that should be there?

File: core/src/modules/DynamicsModule/include/FreeDriftDynamics.hpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            double d = 25000; // 25 km in metres
+            // Fill the fake coordinate array
+            for (size_t j = 0; j < ModelArray::size(ModelArray::Dimension::YVERTEX); ++j) {
+                for (size_t i = 0; i < ModelArray::size(ModelArray::Dimension::XVERTEX); ++i) {
+                    fake25kmCoords.components({i, j})[0] = d * i;
+                    fake25kmCoords.components({i, j})[1] = d * j;
+                }
+            }
+            coords = fake25kmCoords;
+            // End of code to be removed
```
Comment (einola) [inline core/src/modules/DynamicsModule/include/FreeDriftDynamics.hpp:73]: Can we remove this now?

File: dynamics/src/include/FreeDriftDynamicsKernel.hpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    const double cosOceanAngle = cos(radians * params.ocean_turning_angle);
+    const double sinOceanAngle = sin(radians * params.ocean_turning_angle);
+    const double NansenNumber = sqrt(params.F_atm / params.F_ocean);
+
+    void updateMomentum(const TimestepTime& tst) override
+    {
+#pragma omp parallel for
+        for (int i = 0; i < u.rows(); ++i) {
+            // Ice velocity is equal to ocean velocity
```
Comment (einola) [inline dynamics/src/include/FreeDriftDynamicsKernel.hpp:63]: That comment is not correct anymore

```suggestion
            // "Classic free drift", assuming zero atmospheric turning angle.
```

**Review by einola — CHANGES_REQUESTED** [review]
The math looks fine, but there's still a reference to the "fake" spherical coordinates. This can be removed, right?

File: core/src/modules/DynamicsModule/include/FreeDriftDynamics.hpp (hunk 1 above)
Comment (timspainNERSC) [reply core/src/modules/DynamicsModule/include/FreeDriftDynamics.hpp]: Yes, done.

File: dynamics/src/include/FreeDriftDynamicsKernel.hpp (hunk 2 above)
Comment (timspainNERSC) [reply dynamics/src/include/FreeDriftDynamicsKernel.hpp]: Eh, close enough.

**Review by einola — APPROVED** [review]

File: core/src/modules/DynamicsModule/module.cfg [hunk 3]
```diff
@@ -21,3 +21,8 @@ has_help = false
 file_prefix = BBMDynamics
 description = Discontinuous Galerkin dynamics with BBM rheology.
 has_help = false
+
+[Nextsim::FreeDriftDynamics]
+file_prefix = FreeDriftDynamics
+description = Free drift dynamics.
+has_help = false
```
Comment (MarionBWeinzierl) [inline core/src/modules/DynamicsModule/module.cfg:28]: Just out of curiosity: What is the "has help" field for, what does it define/do?

**Review by MarionBWeinzierl — APPROVED** [review]
This looks sensible to me.

File: core/src/modules/DynamicsModule/module.cfg (hunk 3 above)
Comment (timspainNERSC) [reply core/src/modules/DynamicsModule/module.cfg:28]: Within a Module, there is a function that fetches the help text from any implementation that has configuration options (this is the text that is shown when running `./nextsim --help-config`). When I was writing the Module classes by hand, I had to choose which of the implementation sub-classes were interrogated to get help text. This field in the module configuration determines the same thing for the scripts that are building the `Module` classes.
There may be a better way to do this, but that should be another issue & PR.

**Pull request merged**